typescript/src/schema/proto.ts

```diff
@@ -526,55 +527,52 @@ function augmentService(
 
   // Build a list of resources referenced by this service
   const uniqueResources: { [name: string]: ResourceDescriptor } = {};
+  // Copy all resources in resourceDatabase to uniqueResources
+  const allPatterns = resourceDatabase.patterns;
+  for (const pattern of Object.keys(allPatterns)) {
+    const resource = allPatterns[pattern];
+    uniqueResources[resource.name] = resource;
+  }
+
+  // Copy all resources definination which are referenced into unique resources map.
   for (const property of Object.keys(messages)) {
     const errorLocation = `service ${service.name} message ${property}`;
-    // take the option['.google.api.resource'] of the message as resource, add it to resourceDatabase id it's not there.
-    const descriptorProto = messages[property];
-    if (
-      descriptorProto.options &&
-      descriptorProto.options['.google.api.resource']
-    ) {
-      const resource = descriptorProto.options['.google.api.resource'];
-      if (!resourceDatabase.getResourceByType(resource.type)) {
-        resourceDatabase.registerResource(resource);
-        const registeredResource = resourceDatabase.getResourceByType(
-          resource.type
-        )!;
-        uniqueResources[registeredResource.name] = registeredResource;
-      }
-    }
     for (const fieldDescriptor of messages[property].field ?? []) {
       // note: ResourceDatabase can accept `undefined` values, so we happily use optional chaining here.
       const resourceReference =
         fieldDescriptor.options?.['.google.api.resourceReference'];
-
       // 1. If this resource reference has .child_type, figure out if we have any known parent resources.
-      const parentResources = resourceDatabase.getParentResourcesByChildType(
+      const parentResources = resourceDefinitionDatabase.getParentResourcesByChildType(
         resourceReference?.childType,
         errorLocation
       );
       parentResources.map(
         resource => (uniqueResources[resource.name] = resource)
       );
 
-      // 2. If this resource reference has .type, we should have a known resource with this type.
-      const resourceByType = resourceDatabase.getResourceByType(
-        resourceReference?.type,
-        errorLocation
+      // 2. If this resource reference has .type, we should have a known resource with this type, check two maps.
+      let resourceByType = resourceDefinitionDatabase.getResourceByType(
+        resourceReference?.type
       );
+      resourceByType = resourceByType
+        ? resourceByType
```

Comment:
This can now be written as `??`:  

```ts
resourceByType = resourceByType ?? resourceDatabase.getResourceByType(....);
```

New language feature!

Reply:
👍